### PR TITLE
ev3dev: use file registry to reduce file opening costs

### PR DESCRIPTION
**Pull request by kortschak**
This is a counter proposal for #94.

It's not exactly how I'd like it to be, but after several hours dealing with FUSE-based kernel hangs I think it's the best that I'm likely to get to for a while.

Please take a look.

/cc @FlorianRaediker

Closes #94.

### CODE REVIEW

File: ev3dev.go [hunk 1]
```diff
@@ -633,3 +637,89 @@ func setAttributeOf(d Device, attr, data string) error {
...
+		// 128 bytes was suggested in ev3go/ev3dev#93,
+		// but this fails with the LED trigger files.
+		// We log if there is no error since ReadAt
+		// will always return an error if n is less
+		// than len(buf). So we catch all the cases
+		// where the file is longer, with a small number
+		// of false positives where the file is exactly
+		// the length of the buffer. Bump the length
+		// of the buffer when that happens.
+		log.Printf("ev3dev: buffer too short for %s: falling back to ioutil.ReadFile", path)
```
Comment (dlech) [inline ev3dev.go:699]: Couldn't this be an error other than buffer size as well?

**Review by dlech — COMMENTED** [review]
Seems reasonable. It will take some actual use to see how it holds up.

File: ev3dev.go (hunk 1 above)
Comment (kortschak) [reply ev3dev.go:699]: No, it's a non-error. The only way you can have a nil error after a call to the `*os.File` `ReadAt` method is if the buffer is the same length as or shorter than the data being read. All the non-nil error states are handled below.

**Comment by kortschak** [comment]
@FlorianRaediker would you test this branch on your device please?

**Comment by FlorianRaediker** [comment]
It seems to be working well so far. 10,000 value checks take ~7s; using 128 buffer size instead of 256 might be a bit faster (though this is probably negligible).
I also ran a program that uses two Lego large motors and the LineLeader and didn't encounter any problems.
If I do notice any problems in the future, I'll let you know.

**Comment by kortschak** [comment]
> using 128 buffer size instead of 256 might be a bit faster

Yes, for cases where the file is less than 256 bytes. For the files in the LED system, it would be significantly slower.

I'll leave this open for a week and then, baring no issues being raised by @FlorianRaediker, I'll merge.

**Pull request merged**